pkg/auth/providers/saml/saml_client.go

```diff
@@ -326,10 +326,13 @@ func (s *Provider) HandleSamlAssertion(w http.ResponseWriter, r *http.Request, a
 	userID := s.clientState.GetState(r, "Rancher_UserID")
 	if userID != "" && rancherAction == testAndEnableAction {
 		user, err := s.userMGR.SetPrincipalOnCurrentUserByUserID(userID, userPrincipal)
-		if err != nil {
+		if err != nil && user == nil {
 			log.Errorf("SAML: Error setting principal on current user %v", err)
 			http.Redirect(w, r, redirectURL+"errorCode=500", http.StatusFound)
 			return
+		} else if err != nil && user != nil {
+			http.Redirect(w, r, redirectURL+"errorCode=422&errorMsg="+url.QueryEscape(err.Error()), http.StatusFound)
```

Comment:
Did you confirm UI is handling this as expected? It's almost identical to code we already have and `errorMsg` is error ID in there that UI uses to come up with the actual error message and is the actual error message in this case. (Actually glancing through UI code it looks like it will fallback to using errorMsg verbatim if it cannot find it as code). Still, for i18n support we better use code here instead. I'm OK making it a separate enhancement thought and let this one go in as is due to this improvement requiring both UI and backend change.

```
		// UI uses this translation key to get the error message
		http.Redirect(w, r, redirectURL+"errorCode=422&errorMsg="+UITranslationKeyForErrorMessage, http.StatusFound)
```

Reply:
Yes it works with the up, there was not a code for this error message so I sent this for now.  We can open it as a separate issue.